### ISSUE TITLE
Using mold for faster linking

### ISSUE THREAD

**Issue by arbrauns** [zephyrproject-rtos/zephyr#57844]
**Is your enhancement proposal related to a problem? Please describe.**
A significant part of the total build time for a zephyr project can be taken up by the linking of up to three ELF binaries, especially for incremental builds. The [mold linker](https://github.com/rui314/mold) is specifically designed as a significantly faster, modern alternative to existing linkers.

Unfortunately, it is currently targeted mostly at (linux) OS packaging rather than embedded use. As such, there are at least a few outstanding issues that make it unsuitable for the time being, among them:

- No support for linker file [`SECTIONS`](https://github.com/rui314/mold/issues/243)
- No support for `--sort-common=descending`

**Describe the solution you'd like**
The build system automatically uses mold if available, for faster builds without user action.

**Describe alternatives you've considered**
The user can opt-in to using mold using a CMake directive.

**Comment by nashif** [comment]
@tejlmand  @stephanosio  FYI

**Comment by stephanosio** [comment]
I cannot possibly think of a way we can use this linker when it does not have a proper support for the linker scripts; moreover, its author does not seem to be very keen on supporting one either.

I would suggest closing this for now and re-opening it after `mold` becomes a proper linker that supports linker scripts.

**Comment by arbrauns** [comment]
Sure, I wasn't suggesting supporting it in its current state, just wanted to post a "tracking issue" to be referred to in the future.

**Comment by nordicjm** [comment]
> A significant part of the total build time for a zephyr project can be taken up by the linking of up to three ELF binaries, especially for incremental builds.

I mean, really? Are you running on windows? And how slow is slow? Because I've certainly not noticed linking to be slow on ESXi VMs, desktops or laptops on linux.

**Comment by arbrauns** [comment]
> I mean, really? Are you running on windows? And how slow is slow? Because I've certainly not noticed linking to be slow on ESXi VMs, desktops or laptops on linux.

In one of my current projects, touching a single source file causes a subsequent `west build` to take 28 seconds. Actually recompiling the C file only takes a fraction of a second, of course, the rest is linking `zephyr_pre0.elf`, `zephyr_pre1.elf` and `zephyr.elf`.

Linux, i7-11700, project directory on an NVMe SSD.

**Comment by zephyrbot** [comment]
Hi @tejlmand, @evgeniy-paltsev, @nashif, @tejlmand, @abrodkin,

This issue, marked as an Enhancement, was opened a while ago and did not get any traction. It was just assigned to you based on the labels. If you don't consider yourself the right person to address this issue, please re-assing it to the right person.

Please take a moment to review if the issue is still relevant to the project.  If it is, please provide feedback and direction on how to move forward. If it is not, has already been addressed, is a duplicate, or is no longer relevant, please close it with a short comment explaining the reason.


@arbrauns you are also encouraged to help moving this issue forward by providing additional information and confirming this request/issue is still relevant to you.

Thanks!

**Comment by arbrauns** [comment]
I don't think there's anything actionable here yet, this is waiting on improvements in upstream mold.

**Comment by fox1e-dev** [comment]
Were there any updates?
From mold, this discussion seems to be related: https://github.com/rui314/mold/discussions/1261
And this issue: https://github.com/rui314/mold/issues/1055

In the discussion, the author says that embedded support makes sense; however, they would not like to use linker scripts for it.

**Comment by arbrauns** [comment]
I don't think this will move forward until mold supports linker scripts in some form or another. If they do decide to come up with their own variant of linker scripts, as discussed in those issues, zephyr will also need quite a bit of logic to emit that in addition to the "standard" linker scripts supported by all other linkers.
